### PR TITLE
spirv-val: Multiple interface var with same SC

### PR DESCRIPTION
**Pull request by spencer-lunarg**
adds

- VUID-StandaloneSpirv-OpVariable-06673
- VUID-StandaloneSpirv-IncomingRayPayloadKHR-04700
- VUID-StandaloneSpirv-HitAttributeKHR-04702
- VUID-StandaloneSpirv-IncomingCallableDataKHR-04706

### CODE REVIEW

File: source/val/validate_interfaces.cpp [hunk 1]
```diff
@@ -533,6 +533,64 @@ spv_result_t ValidateLocations(ValidationState_t& _,
...
+    auto storage_class = interface_var->GetOperandAs<SpvStorageClass>(2);
+    switch (storage_class) {
+      case SpvStorageClassPushConstant: {
+        if (has_push_constant) {
+          return _.diag(SPV_ERROR_INVALID_DATA, entry_point)
+                 << _.VkErrorID(6673)
+                 << "Entry-point has more than one variable with the "
+                    "PushConstant storage class in the interface";
+        }
+        has_push_constant = true;
```
Comment (alan-baker) [inline source/val/validate_interfaces.cpp:554]: What's Vulkan's stance on duplicate information generally? VUID 6674 says you can't use more than one variable with a push constant storage class per entry point. That covers all module versions. This seems redundant, but only covering SPIR-V 1.4 or later. There is code in validate_decorations.cpp for VUID 6674. To me, this duplication is not great from a maintenance point of view.

File: source/val/validate_interfaces.cpp [hunk 2]
```diff
@@ -533,6 +533,64 @@ spv_result_t ValidateLocations(ValidationState_t& _,
...
+        break;
+      }
+      case SpvStorageClassIncomingRayPayloadKHR: {
+        if (has_ray_payload) {
+          return _.diag(SPV_ERROR_INVALID_DATA, entry_point)
+                 << _.VkErrorID(4700)
+                 << "Entry-point has more than one variable with the "
+                    "IncomingRayPayloadKHR storage class in the interface";
+        }
+        has_ray_payload = true;
```
Comment (alan-baker) [inline source/val/validate_interfaces.cpp:564]: While this is not a duplicate, if there is a change for push constants I wonder if the language should be maintained in a parallel manner.

**Review by alan-baker — APPROVED** [review]

**Comment by cassiebeckley** [comment]
@spencer-lunarg this is causing failures in DXC's test suite (see https://ci.appveyor.com/project/dnovillo/directxshadercompiler/builds/45540244/job/ga5ytxvk7g21bexf), and I'm not sure whether it's a bug in this PR or if DXC is generating invalid code.

The tests are https://github.com/microsoft/DirectXShaderCompiler/blob/main/tools/clang/test/CodeGenSPIRV/vk.layout.shader-record-ext.std430.hlsl and https://github.com/microsoft/DirectXShaderCompiler/blob/main/tools/clang/test/CodeGenSPIRV/raytracing.khr.terminate.hlsl

**Comment by spencer-lunarg** [comment]
@cassiebeckley (I'm no RTX expert) but in `vk.layout.shader-record-ext.std430.hlsl` 

> OpEntryPoint ClosestHitNV %chs1 "chs1" %cbuf %block %P %A %P_0 %A_0

Is the issue that `inout Payload P` is turning into 2 variables in DXC

```swift
OpEntryPoint ClosestHitNV %1 "chs1" %2 %3 %P %P_0

%struct = OpTypeStruct %19
%ptr = OpTypePointer IncomingRayPayloadNV %struct

%P = OpVariable %ptr IncomingRayPayloadNV
%P_0 = OpVariable %ptr IncomingRayPayloadNV
```

and they are pointing to the same location. To my knowledge the "input"/"output" part is not something indexed and should be just a single variable. (So I think this is a DXC bug)

**Pull request merged**